app/src/org/commcare/dalvik/activities/CreatePinActivity.java

```diff
@@ -0,0 +1,220 @@
+package org.commcare.dalvik.activities;
+
+import android.content.DialogInterface;
+import android.content.Intent;
+import android.os.Bundle;
+import android.text.Editable;
+import android.text.TextWatcher;
+import android.util.Log;
+import android.view.Menu;
+import android.view.MenuItem;
+import android.view.View;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import org.commcare.android.database.app.models.UserKeyRecord;
+import org.commcare.android.framework.ManagedUi;
+import org.commcare.android.framework.SessionAwareCommCareActivity;
+import org.commcare.android.framework.UiElement;
+import org.commcare.dalvik.R;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.commcare.dalvik.dialogs.AlertDialogFactory;
+import org.javarosa.core.services.locale.Localization;
+
+
+/**
+ * @author Aliza Stone (astone@dimagi.com)
+ */
+@ManagedUi(R.layout.create_pin_view)
+public class CreatePinActivity extends SessionAwareCommCareActivity<CreatePinActivity> {
+
+    private static final int MENU_REMEMBER_PW_AND_LOGOUT = Menu.FIRST;
+
+    @UiElement(value=R.id.pin_entry)
+    private EditText enterPinBox;
+
+    @UiElement(value=R.id.pin_prompt_text)
+    private TextView promptText;
+
+    @UiElement(value=R.id.pin_cancel_button)
+    private Button cancelButton;
+
+    @UiElement(value=R.id.pin_confirm_button)
+    private Button continueButton;
+
+    private static final String TAG = CreatePinActivity.class.getSimpleName();
+    public static final String CHOSE_REMEMBER_PASSWORD = "chose-remember-password";
+
+    private String unhashedUserPassword;
+    private LoginMode loginMode;
+    private UserKeyRecord userRecord;
+
+    // Indicates whether the user is entering their PIN for the first time, or is confirming it
+    private boolean inConfirmMode;
+    private String firstRoundPin;
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        userRecord = CommCareApplication._().getRecordForCurrentUser();
+        if (userRecord == null) {
+            Log.i(TAG, "Something went wrong in CreatePinActivity. Could not find the current user " +
+                    "record, so just finishing the activity");
+            setResult(RESULT_CANCELED);
+            this.finish();
+            return;
+        }
+
+        loginMode = LoginMode.fromString(
+                getIntent().getStringExtra(LoginActivity.LOGIN_MODE));
+        if (loginMode == LoginMode.PASSWORD) {
+            unhashedUserPassword = getIntent().getStringExtra(LoginActivity.PASSWORD_FROM_LOGIN);
+        } else if (loginMode == LoginMode.PRIMED) {
```

Comment:
I'm having trouble figuring out when the login mode will be `PASSWORD` and when it will be `PRIMED`.